internal/generate/run_test.go

```diff
@@ -128,290 +128,117 @@ func TestRun(t *testing.T) {
 		testfs.AssertEqualDir(t, assertdir, destdir)
 	})
 
-	t.Run("success_export", func(t *testing.T) {
```

Comment:
**suggestion :** Apply the following change to:
- Ensure the next option doesn't force the developer to add it in the test case struct
- Doesn't get weird behavior on `tc.SamePackage` with overridding every previous defined variable

```go
type testCase struct {
		generate.CLIOptions

		DirName     string
		SamePackage bool
	}
	for _, tc := range []testCase{
		{
			DirName: "success_channels",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Chan"},
			},
		},
		{
			DirName: "success_export",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Export"},
			},
		},
		{
			DirName: "success_funcs",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Func"},
			},
		},
		{
			DirName: "success_interface",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Interface", "InterfaceNoFields"},
			},
		},
		{
			DirName: "success_maps",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Map"},
			},
		},
		{
			DirName: "success_module_replace",
			CLIOptions: generate.CLIOptions{
				File:    "module::github.com/sirupsen/logrus/hooks/test/test.go",
				Structs: []string{"Hook"},
			},
		},
		{
			DirName: "success_module_root",
			CLIOptions: generate.CLIOptions{
				File:    "module::github.com/go-playground/validator/v10/errors.go",
				Structs: []string{"InvalidValidationError"},
			},
		},
		{
			DirName: "success_module_subdir",
			CLIOptions: generate.CLIOptions{
				File:    "module::github.com/sirupsen/logrus/hooks/test/test.go",
				Structs: []string{"Hook"},
			},
		},
		{
			DirName: "success_naming",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Naming"},
			},
		},
		{
			DirName: "success_root_gomod",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"RootType"},
			},
		},
		{
			DirName: "success_same_package",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"SamePackage", "unexportedType"},
			},
			SamePackage: true,
		},
		{
			DirName: "success_same_package_prefix",
			CLIOptions: generate.CLIOptions{
				Structs:      []string{"unexportedTypePrefix"},
				SetterPrefix: "Set",
			},
			SamePackage: true,
		},
		{
			DirName: "success_slices",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"ArrayAndSlice"},
			},
		},
		{
			DirName: "success_struct",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Struct", "StructNoFields"},
			},
		},
		{
			DirName: "success_with_options",
			CLIOptions: generate.CLIOptions{
				Structs:      []string{"Options", "Empty"},
				ValidateFunc: "Validate",
			},
		},
		{
			DirName: "success_return_copy",
			CLIOptions: generate.CLIOptions{
				Structs:    []string{"ReturnCopy"},
				ReturnCopy: true,
			},
		},
		{
			DirName: "success_generics",
			CLIOptions: generate.CLIOptions{
				Structs: []string{"Struct"},
			},
		},
	} {
		t.Run(tc.DirName, func(t *testing.T) {
			// Arrange
			assertdir, destdir := func() (string, string) {
				destdir := t.TempDir()
				assertdir := filepath.Join(testdata, tc.DirName)

				if tc.File == "" {
					tc.File = filepath.Join(assertdir, "types.go")
				}

				if !tc.SamePackage {
					destdir := filepath.Join(destdir, "builders")
					assertdir := filepath.Join(assertdir, "builders")
					return assertdir, destdir
				}

				src := tc.File
				tc.File = filepath.Join(destdir, "types.go")
				require.NoError(t, filesystem.CopyFile(src, tc.File))
				return assertdir, destdir
			}()
			tc.CLIOptions.Destdir = destdir

			// Act
			err := generate.Run(pwd, tc.CLIOptions)

			// Assert
			assert.NoError(t, err)
			testfs.AssertEqualDir(t, assertdir, destdir)
		})
	}
```

It stills feel weird anyway, I'm not a fan of test loop, and in this case I'm shared on the application because we get condition in the test `Arrange` part, adding as such intelligence in tests.

In my opinion, tests loop are also complicated because specifically in vscode it's pain to debug a case only ...

But anyway, I agree there was a lot of duplicated code.

Reply:
Done